Maxim belyj

MaximBelyj opened the pull request:
Наисал класс DAO, в нем написал 2 метода, чтобы добавить животного в БД и чтобы достать, но пока не разобрался с аннотациями

MaximBelyj committed:
Add to ignore some files

MaximBelyj committed:
Create DAO, and database

MaximBelyj committed:
Merge remote-tracking branch 'origin/MaximBelyj' into MaximBelyj

MaximBelyj committed:
Создал DAO, + написал 2 метода, чтобы можно было добалять животных и показывать уже имеющихся

MaximBelyj committed:
add abstract

Alexander091 commented on trainings/Maxim/src/main/java/com/zooshop/DAO.java:
```diff
@@ -0,0 +1,65 @@
+    public DAO() { // конструктор с установкой соединение с БД
+
+        try {
+            Class.forName("org.postgresql.Driver"); //проверяем наличие драйвера
```
Не имеет смысла отлавливать такие ошибки.
и всё из-за того, что ты не сможешь их обработать внутри программы.

Alexander091 commented on trainings/Maxim/src/main/java/com/zooshop/DAO.java:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    public void getAll() { // получаем все строки
+        try{
```
Перечитай ещё раз принцип SOLID, а именно "The Single Responsibility Principle"
Это означает, что один класс/метод должен выполнять одну функцию. Судя по названию, он должен возвращать все объекты.
Что будет дальше с этими объектами нас не должно интересовать.
т.е операции:
System.out.println(object_id + " " + parent_id + " " + object_type_id + " " + name);
тут не должно быть

Alexander091 commented on .gitignore:
```diff
@@ -10,6 +10,10 @@
 *.idea
 *out
 *.iml
+*.xml
```
xml ты зря добавил в ignore

Alexander091 commented on trainings/Maxim/src/main/java/com/zooshop/DAO.java:
```diff
@@ -0,0 +1,65 @@
+
+    }
+
+    public void setAnimal(int obj_id, int par_id, int obj_type_id, String name){ //добавляем в таблицу
```
Название не слишком подходящие, скорее save вместо set

Alexander091 commented on trainings/Maxim/src/main/java/com/zooshop/DAO.java:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    public void setAnimal(int obj_id, int par_id, int obj_type_id, String name){ //добавляем в таблицу
+        String query1 = "INSERT INTO objects (object_id, parent_id, object_type_id, name) VALUES ('"+obj_id+"','"+par_id+"','"+obj_type_id+"', '"+name+"');";
```
Запросы должны быть параметризованы, посмотри в сторону PreparedStatement

Alexander091 commented on trainings/Maxim/src/main/java/com/zooshop/Shop.java:
```diff
@@ -1,23 +1,28 @@
 
-        info.sayWelcome();
+        DAO d = new DAO();
+        d.setAnimal(4, 1, 2, "cat3");
```
Твой магазин не должен знать о том как хранится животное, это не его ответственность. 
Магазин управляет животными. 
Как хранятся эти животные определяет твой DAO. 
Более того, у тебя есть сервис, который достаёт животных, так пускай этот самый сервис использует DAO для получения информации о животных

MaximBelyj committed:
убрал .xml, переписал ShopService,  убрал DAO в ShopService, изменил название setAnimal на saveAnimal, getAll на printAll, ошибку типа ClassNotFound, где проверяется наличие драйвера всёравно отлавливаю потому, что иначе ошибка при использовании объектов DAO в классе ShopService

MaximBelyj committed:
убрал .xml, переписал ShopService,  убрал DAO в ShopService, изменил название setAnimal на saveAnimal, getAll на printAll, ошибку типа ClassNotFound, где проверяется наличие драйвера всёравно отлавливаю потому, что иначе ошибка при использовании объектов DAO в классе ShopService, добавил метод закрыть магазин(закрыть соединение)

MaximBelyj committed:
Переписал с PrepearedStatement DAO, соответсвенно изменил вызов методов в классах ShopService и Shop. + Сделал так , чтобы метод getAllObjects не печатал выво строк, эту функцию я сделал в классе ShopService, НО! я не придумал ничего лучше, чем написать это через Map, тоесть я получаю строки из таблицы, запихиваю их в Map и потом вывожу это на экран .

Alexander091 commented on trainings/Maxim/src/main/java/com/zooshop/DAO.java:
```diff
@@ -0,0 +1,79 @@
+        }
+    }
+
+    public Map getAllObjects() { // получаем все строки
```
Во-первых, я скинул тебе пример того, как получать сущности.
Во-вторых, если ты использует списки, мапы, то используй их типизированные варианты.
Иначе, чтобы получить данные из такой мапы, тебе придётся устраивать приведение типов, а это чревато ошибками и трудностями в отладке

Alexander091 commented on trainings/Maxim/src/main/java/com/zooshop/DAO.java:
```diff
@@ -0,0 +1,79 @@
+        return map;
+    }
+
+    public void saveAnimal(int obj_id, int par_id, int obj_type_id, String name){ //добавляем в таблицу
```
верблюжийСтильНужноСоблюдать

MaximBelyj committed:
Кое-что по мелочи исправил

MaximBelyj committed:
Кое-что по мелочи исправил

MaximBelyj committed:
Дописал чтобы аннотация могла применяться только к полям, а также чтобы была доступна при нпследовании,
убрал из конструкторов параметры, написал новый метод который добавляет животное в таблицы objects и params

MaximBelyj committed:
Сделал добавление сущностей + удаление сущностей, лень писать, ну много изменений сделал

Alexander091 commented on trainings/Maxim/src/main/java/com/DAO/DAO.java:
```diff
@@ -0,0 +1,195 @@
+    public <T> void save(T obj) throws IllegalAccessException {
+        Class cl = obj.getClass();
+        Class cl2 = cl.getSuperclass();
+        if (obj instanceof Dog){
```
Никакой информации о кошках и собаках тут не должно быть.
А если у нас будет 100 различных типов, ты каждый планируешь проверять?
У нас есть для этого аннотация ObjectType

Alexander091 commented on trainings/Maxim/src/main/java/com/DAO/DAO.java:
```diff
@@ -0,0 +1,195 @@
+                    preparedStatement = connection.prepareStatement(QUERY_INSERT_VALUE);
+                    switch (attr.value()) {
+                        case 1:
+                            preparedStatement.setInt(1, 1);
```
А почему ты подставляешь жёстко зашитые значения? 
Нужно их брать с объекта, который передался на вход

Alexander091 commented on trainings/Maxim/src/main/java/com/DAO/DAO.java:
```diff
@@ -0,0 +1,195 @@
+                Attribute attr = field.getAnnotation(Attribute.class);
+                try {
+                    preparedStatement = connection.prepareStatement(QUERY_INSERT_VALUE);
+                    switch (attr.value()) {
```
И зачем тут switch? какую цель он выполняет?

Alexander091 commented on trainings/Maxim/src/main/java/com/DAO/DAO.java:
```diff
@@ -0,0 +1,195 @@
+                            break;
+                        case 4:
+                            preparedStatement.setInt(1, 4);
+                            preparedStatement.setString(2, (String) field.get(obj));
```
кусок кода повторяется повсюду... в таких случаях это нужно выносить в отдельную функцию.
Дублирование кода это зло, только увеличивает сложность сопровождения

Alexander091 commented on trainings/Maxim/src/main/java/com/DAO/DAO.java:
```diff
@@ -0,0 +1,195 @@
+                Attribute attrAnnotation = field.getAnnotation(Attribute.class);
+                int attrId = attrAnnotation.value();
+                System.out.println(attrId);
+                for (Map.Entry<Long, String> entry : resultSetObjectValues.entrySet()) {
```
используй функцию https://docs.oracle.com/javase/7/docs/api/java/util/Map.html#get(java.lang.Object)
вместо цикла

Alexander091 commented on trainings/Maxim/src/main/java/com/zoo/Entity.java:
```diff
@@ -0,0 +1,35 @@
+    private int id;   //id сущности
+    private String name;    //имя сущности
+    @Attribute(7)
+    private int type;//тип сущности
```
что это такое "Тип сущности"?

Alexander091 commented on trainings/Maxim/src/main/java/com/zooshop/Shop.java:
```diff
@@ -0,0 +1,106 @@
+
+        Animal cat = new Cat();
+        Animal dog = new Dog();                  //TODO Dependency Injecion
+        ShopService sh = new ShopService();     //TODO добавить инициализацию, в DAO и ShopService,
```
Спринговые бины. Не забудь про них

MaximBelyj committed:
...

MaximBelyj committed:
ignore .xml

Alexander091 commented on trainings/Maxim/src/main/java/com/DAO/DAO.java:
```diff
@@ -0,0 +1,195 @@
+    private final static String PASSWORD = "123";
+    private final static String QUERY_SELECT = "SELECT  o.object_id, o.name, p.value FROM  objects o JOIN params p ON o.object_id = p.object_id WHERE o.object_id = ?;"; //запрос
+    private final static String QUERY_INSERT = "INSERT INTO objects (parent_id, object_type_id, name) VALUES (?, ?, ?)";
+    private final static String QUERY_INSERT_VALUE = "INSERT INTO params (attr_id, value) VALUES (?, ?)";
```
что-то мне подсказывает, что в параметрах должен быть ещё object_id

MaximBelyj committed:
Убрал (obj instanceof Dog) вместо этого сделал определение типа по значению аннотации, описал предназначение switch в DAO,  убрал тип сущности

MaximBelyj committed:
Переписал Entity так, чтобы он хранил параметры, котроые в таблице objects

MaximBelyj committed:
Переписал Entity так, чтобы он хранил параметры, котроые в таблице objects

MaximBelyj committed:
Также переписал Animal

MaximBelyj committed:
Переписал классы Cat и Dog

MaximBelyj committed:
Внёс изменения + добавил комментов

MaximBelyj committed:
Написал методы по добавлению и удалению сущности из/в таблицу/ы

MaximBelyj committed:
Дописал все методы для DAO, также немного модифицировал класс Dialogue

MaximBelyj committed:
Дописал все методы для DAO, также немного модифицировал класс Dialogue

MaximBelyj committed:
Убрал запрос на удаление из таблицы params вместо этого сделал каскадное удаление в свойствах внешенго ключа object_id в таблице params

MaximBelyj committed:
Внёс ещё изменений - добавил поле род в Animal, ввёл два имени - просто имя животного которое мы задаём и имя типа животного, переписал класс диалог

MaximBelyj committed:
Добавил wildfly

MaximBelyj committed:
Добавил Wildfly

MaximBelyj committed:
a

MaximBelyj committed:
a

MaximBelyj committed:
Ещё одна программа, написал сервлеты, создал WAR моего проекта, задеплоил на сервер wildfly.

MaximBelyj committed:
добавил логгирование + spring модуль

MaximBelyj committed:
добавил логгирование + spring модуль

MaximBelyj committed:
добавил логгирование + spring модуль

MaximBelyj committed:
qwerty

MaximBelyj committed:
qwerty

MaximBelyj committed:
qwerty

MaximBelyj committed:
дописал удаление и кое что другое
